cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,449 @@
+      super(Opcodes.ASM5, cw);
+      this.className = className;
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
+      }
+
+      final boolean[] hasAuthEnforce = new boolean[1];
+      final AuthEnforce[] authEnforceAnnotation = new AuthEnforce[1];
+      final Map<String, Integer> paramAnnotation = new HashMap<>();
+
+      // Visit the annotations of the method to determine if it has AuthEnforce annotation. If it does then collect
+      // all the AuthEnforce annotation details through AnnotationVisitor and also sets a boolean flag hasEnforce
+      // which is used later in visitParameterAnnotation to visit the annotations on parameters of this method only
+      // if it had AuthEnforce annotation.
+      mv = new MethodVisitor(Opcodes.ASM5, mv) {
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+          if (visible) {
+            // if the annotation is present then visit the annotation
+            if (AuthEnforce.class.getName().equals(Type.getType(desc).getClassName())) {
+              // set the flag which will be used to decide whether to visit parameter annotation or not
+              hasAuthEnforce[0] = true;
+              final Class[] enforceOn = new Class[1];
+              final List<String> entities = new ArrayList<>();
+              final Set<Action> actions = new HashSet<>();
+
+              // We found AuthEnforce annotation so visit the annotation to collect all the specified details
+              // See AnnotationVisitor documentation to know the order in which the below overridden methods are
+              // called and their responsibilities.
+              annotationVisitor = new AnnotationVisitor(Opcodes.ASM5, annotationVisitor) {
+                // gets the entity id class on which enforcement has to performed.
+                @Override
+                public void visit(String name, Object value) {
+                  try {
+                    enforceOn[0] = Class.forName(((Type) value).getClassName());
```

Comment:
You don't need to load the `Class`, right?? You just need the asm `Type`, isn't it?